pdx-extract/app/main.py

```diff
@@ -0,0 +1,174 @@
+    }
+    return requests.post(
+        urls.SEARCH_URL,
+        headers={"referer": urls.INIT_URL},
+        cookies=cookies,
+        params=payload,
+    )
+
+
+def get_schedule_json(subject, term, unique_session_id, cookies):
+    """Gets JSON representation of the subject for the specified term.
+
+    :param subject: The subject in question.
+    :param term: The term in question.
+    :param unique_session_id: Unique session id generated by the page which
+                              allows authentication.
+    :param cookies: Cookies of the previous requests.
+    """
+
+    payload = {
+        "txt_subject": subject["code"],
+        "txt_term": term["code"],
+        "startDatepicker": "",
+        "endDatepicker": "",
+        "uniqueSessionId": unique_session_id,
+        "pageOffset": "0",
+        "pageMaxSize": "100",
+        "sortColumn": "subjectDescription",
+        "sortDirection": "asc",
+    }
+    res = requests.get(
+        urls.SCHEDULE_URL,
+        headers={"referer": urls.CLASS_URL},
+        cookies=cookies,
+        params=payload,
+    )
+
+    if res.ok:
+        return res.json()
+
+    return None
+
+
+def get_subjects(cookies, unique_session_id, term_date):
+    """Gets the subjects that are available for a particular term.
+
+    :param cookies: Cookies needed to authenticate the request.
+    :param unique_session_id: Parameter needed to authenticate the request.
+    :param term_date: term Where the subjects will be searched for.
+    :returns: JSON with list of subjects
+    """
+
+    payload = {
+        "uniqueSessionId": unique_session_id,
+        "dataType": "json",
+        "searchTerm": "",
+        "term": term_date,
+        "offset": "1",
+        "max": config.MAX_SUBJECTS,
+        # Query string params expect a timestamp with extra 3 digits
+        "_:": str(int(time.time() * 1000)),
+    }
+    res = requests.get(urls.SUBJECTS_URL, cookies=cookies, params=payload)
+
+    if res.ok:
+        return res.json()
+
+    return None
+
+
+def get_terms(cookies, unique_session_id):
+    """Gets JSON with list of terms in the form {code : description}
+
+    :param cookies: Cookies needed to authenticate the request
+    :param unique_session_id: Parameter needed to authenticate the request
+    :returns: JSON with list of the terms
+    """
+
+    payload = {
+        "uniqueSessionId": unique_session_id,
+        "dataType": "json",
+        "searchTerm": "",
+        "offset": "1",
+        "max": config.MAX_TERMS,
+    }
+    res = requests.get(urls.TERMS_URL, cookies=cookies, params=payload)
+
+    if res.ok:
+        return res.json()
+
+    logger.info("No terms were found.")
+    return None
+
+
+async def run():
+    browser = await pyppeteer.initialize()
+    page = await pyppeteer.get_page(browser)
+    session_id, unique_session_id = await pyppeteer.get_tokens(browser)
+
+    if None in (session_id, unique_session_id):
+        browser.close()
+        return
```

Comment:
Is there any value in having something logged here? Do you have access to the plumbing to do useful logging?

Reply:
Yes, and yes! I'm using structured logging to make the messages parseable by a 3rd party logging application. I'll add an error-level message here.